src/scalarstats.jl

```diff
@@ -261,36 +261,92 @@ realXcY(x::Real, y::Real) = x*y
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
+
+# References
+
+Carl-Erik Särndal, Bengt Swensson, Jan Wretman (1992). Model Assisted Survey Sampling.
+New York: Springer. pp. 51-53.
+"""
+sem(x, weights::AbstractWeights; mean=nothing) = _sem(x, weights, mean)
+sem(x; mean=nothing) = _sem(x, mean)
+
+function _sem(x, mean)
+    n = 0
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        variance = var(x; mean=mean, corrected=true)
+    else
+        sse = real(zero(mean))
+        for element in x
+            n += 1
+            sse += abs2(element - mean)
+        end
+        variance = sse / (n - 1)
     end
-    count = 1
-    value, state = y
-    y = iterate(x, state)
-    # Use Welford algorithm as seen in (among other places)
-    # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
-    M = value / 1
-    S = real(zero(M))
-    while y !== nothing
+    return sqrt(variance / n)
+end
+
+function _sem(x, ::Nothing)
+    n = 0
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        variance = var(x; corrected=true)
+    else
+        y = iterate(x)
         value, state = y
-        y = iterate(x, state)
-        count += 1
-        new_M = M + (value - M) / count
-        S = S + realXcY(value - M, value - new_M)
-        M = new_M
+        # Use Welford algorithm as seen in (among other places)
+        # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
+        mean = value / 1
+        sse = real(zero(mean))
+        while y !== nothing
+            value, state = y
+            y = iterate(x, state)
+            n += 1
+            new_mean = mean + (value - mean) / n
+            sse += realXcY(value - mean, value - new_mean)
+            mean = new_mean
+        end
+        variance = sse / (n - 1)
     end
-    var = S / (count - 1)
-    return sqrt(var/count)
+    return sqrt(variance / n)
+end
+
+_sem(x::AbstractArray, mean) = sqrt(var(x; mean=mean, corrected=true) / length(x))
+_sem(x::AbstractArray, ::Nothing) = sqrt(var(x; corrected=true) / length(x))
+
+function _sem(x::AbstractArray, weights::UnitWeights, mean)
+    if length(x) ≠ length(weights)
+        throw(DimensionMismatch("array and weights do not have the same length"))
+    end
+    return sem(x; mean=mean)
+end
+
+# Weighted methods for the above
+_sem(x::AbstractArray, weights::FrequencyWeights, mean) =
+    sqrt(var(x, weights; mean=mean, corrected=true) / sum(weights))
+
+_sem(x::AbstractArray, weights::ProbabilityWeights, ::Nothing) = 
+    _sem(x, weights, mean(x, weights))
+
+function _sem(x::AbstractArray, weights::ProbabilityWeights, mean)
+    # sum of squared errors = sse
+    sse = sum(Broadcast.instantiate(Broadcast.broadcasted(x, weights) do x_i, w
+        return abs2(w * (x_i - mean))
+    end))
```

Comment:
```suggestion
function sem(x::AbstractArray, weights::ProbabilityWeights; mean=nothing)
    _mean = mean === nothing ? Statistics.mean(x, weights) : mean
    # sum of squared errors = sse
    sse = sum(Broadcast.instantiate(Broadcast.broadcasted(x, weights) do x_i, w
        return abs2(w * (x_i - _mean))
    end))
```

Reply:
What's the advantage of this? I thought breaking it up into two methods (one where a mean is passed and one without) made it more clear.

Reply:
Well if that allows dropping the internal `_sem` function, the overall result is simpler.